### ISSUE TITLE
`ProgramDirectory` appends to package.path; should prepend instead?

### ISSUE THREAD

**Issue by michaellenaghan** [jart/cosmopolitan#862]
Add this to `.init.lua`:

```lua
Log(kLogVerbose, " *** package.path: " .. package.path)
```

Now run `redbean -D something`:

```
V2023-07-21T15:23:03+000226:/.init.lua:3:redbean-unsecure-2:13611]  *** package.path: /zip/.lua/?.lua;/zip/.lua/?/init.lua
```

The directory wasn't added to `package.path`.

Now add this to `.init.lua`, before the `Log()` line:

```lua
ProgramDirectory("something")
```

The directory still wasn't added.

The problem is in `UpdateLuaPath()`:

https://github.com/jart/cosmopolitan/blob/18536950b3db37da45641bc5805d82ea6056d283/tool/net/redbean.c#L1087-L1100

I'll submit a pull request once I work out how to do a build. In the meantime, there's a workaround of sorts: set `package.path` manually in `.init.lua`, like this:

```lua
package.path = "something/.lua/?.lua;something/.lua/?/init.lua;" .. package.path
```

===

On a related note, `LuaStart()` calls `GetDefaultLuaPath()`:

https://github.com/jart/cosmopolitan/blob/18536950b3db37da45641bc5805d82ea6056d283/tool/net/redbean.c#L5353

`g_lua_path_default` is never actually used; both it and `GetDefaultLuaPath()` can safely be removed.

**Comment by pkulchenko** [comment]
@michaellenaghan, are you sure you're using the latest master branch. This works for me:

```
> redbean.com -D backup -e "print(package.path); unix.exit()"
I2023-07-21T16:05:58.182708:tool/net/redbean.c:1111:redbean:12636] (cfg) program directory: backup
/zip/.lua/?.lua;/zip/.lua/?/init.lua;backup/.lua/?.lua;backup/.lua/?/init.lua
```

It also produces exactly the same output with `redbean.com -e "ProgramDirectory('backup'); print(package.path); unix.exit()"` (assuming `backup` directory exists), so it all appears to work as designed. What version are you using?

> g_lua_path_default is never actually used; both it and GetDefaultLuaPath() can safely be removed.

It's used in lua/loadlib.c to set the initial `package.path` value.

**Comment by michaellenaghan** [comment]
Ah.

I'm using 2.2, which I downloaded from [here](https://redbean.dev/) two days ago. But it looks like a) 2.2 was from last Nov and b) `ProgramDirectory()` was quite different last Nov:

https://github.com/jart/cosmopolitan/blob/5e60e5ad107f0b32d16263ef02dc5090861dc664/tool/net/redbean.c#L1069-L1080

There's no call to `UpdateLuaPath()`. In fact, there's no `UpdateLuaPath()`!

Hmmm, I guess I'll *really* have to work out how to do a build. :-)

On this:

> It's used in lua/loadlib.c to set the initial package.path value.

there might still be a small issue:

 * `RedBean()` calls `LuaStart()`
   * `LuaStart()` calls `GetDefaultLuaPath()`
     * `GetDefaultLuaPath()` iterates over `stagedir`s
 * `RedBean()` calls `GetOpts()`
   * `GetOpts()` calls `ProgramDirectory()`
     * `ProgramDirectory()` calls `AddString(&stagedirs, s, n)`
     * `ProgramDirectory()` calls `UpdateLuaPath()`

It seems pointless — "misleading" might be better? — for `GetDefaultLuaPath()` to iterate over `stagedir`s, because `ProgramDirectory()` is the only thing that updates them, and it hasn't been called yet.

In other words, what we're seeing in your output is a result of the `ProgramDirectory()` call, not the `GetDefaultLuaPath()` call.

Here, maybe, is a bigger issue:

```
/zip/.lua/?.lua;/zip/.lua/?/init.lua;backup/.lua/?.lua;backup/.lua/?/init.lua
```

`UpdateLuaPath()` is *appending* `stagedir`s rather than *prepending* them — so files in the zip will take precedence over files in the `stagedir`s. I think that's wrong?

If you agree, I can close this issue and open a new one, or change the description of this one?

Thanks very much for your follow up!

P.S. The appending is happening here, in `UpdateLuaPath()`:

https://github.com/jart/cosmopolitan/blob/18536950b3db37da45641bc5805d82ea6056d283/tool/net/redbean.c#L1094

**Comment by pkulchenko** [comment]
> UpdateLuaPath() is appending stagedirs rather than prepending them — so files in the zip will take precedence over files in the stagedirs. I think that's wrong?

Not sure; it's consistent with how `stagedirs` behave in other aspects. It only matters if there is a conflict with Lua modules and only for `require`/`dofile`, as `LoadAsset` actually checks for a file before it checks for a zip asset.

Is there a use case when you'd prefer the module to be loaded from the file system instead of the internal archive? hot reloading?

**Comment by michaellenaghan** [comment]
> Not sure; it's consistent with how `stagedirs` behave in other aspects.

I’m not sure how to reconcile that sentence with the ones that follow? I think that everything else calls `GetAsset()`, and `GetAsset()` calls `GetAssetFile()` before `GetAssetZip()` — so isn’t it *in*consistent with everything else?

https://github.com/jart/cosmopolitan/blob/18536950b3db37da45641bc5805d82ea6056d283/tool/net/redbean.c#L2263-L2264

> Is there a use case when you'd prefer the module to be loaded from the file system instead of the internal archive? hot reloading?

I’m trying to figure out a good workflow for fullmoon development. Here I was exploring the idea of incrementally writing code without constantly having to re-add changed files to the zip. (It’s not really a hot reloading scenario because I’m running redbean in the foreground, and redbean doesn’t hot reload in the foreground.)

Given that I have to start and stop redbean anyway, maybe re-adding isn’t so bad? I’m only a day or two into this!

**Comment by pkulchenko** [comment]
> I think that everything else calls GetAsset(), and GetAsset() calls GetAssetFile() before GetAssetZip() — so isn’t it inconsistent with everything else?

Yes, I understand this, but I was referring specifically to `stagedirs`. One issue with prepending is that it causes the order of the directories to be reversed, which is not ideal, as it would only apply to `require`/`dofile`, but not to other cases, like loading assets.

I'm not totally happy about the paths being appended, but still hesitant to prepend, unless we can find some way to prepend the paths, but do it in a way that preserves the order. I'll try to come up with a patch over the weekend.

Reloading required modules is not straightforward, as they still need to be purged from the `package.loaded` table.

> good workflow for fullmoon development. Here I was exploring the idea of incrementally writing code without constantly having to re-add changed files to the zip. (It’s not really a hot reloading scenario because I’m running redbean in the foreground, and redbean doesn’t hot reload in the foreground.)

I have the same preference. In fact I submitted #851, which would allow to write a handler to load assets and to reload modules when a certain signal is received. You can actually send SIGUSR1 to as redbean to rescan its assets, which would trigger the hook added in #851. I'd still need to add its handling to fullmoon when it's available in redbean.

**Comment by michaellenaghan** [comment]
> Yes, I understand this, but I was referring specifically to `stagedirs`. One issue with prepending is that it causes the order of the directories to be reversed, which is not ideal, as it would only apply to `require`/`dofile`, but not to other cases, like loading assets.

Well, `stagedir`s don't apply to *anything* in Lua that doesn't use `package.path`, so whether they're appended or prepended doesn't really make a difference from that perspective.

But I get your bigger point. It's taken me some time and effort to work out the right mental model for what the `stagedir`s do/don't do. At this point my mental model is something like "they affect things that redbean does directly, but nothing that it does indirectly." 

My initial mental model was something like "`stagedir`s + zip are a single virtual filesystem." I have no issue with the way it *actually* works, it's just that it wasn't obvious — to me, at least. Maybe it's just a documentation thing?

> I have the same preference. In fact I submitted https://github.com/jart/cosmopolitan/pull/851, which would allow to write a handler to load assets and to reload modules when a certain signal is received.

I can only think of one reliable way to hot reload Lua. I'll explain that below, but first: after pondering this overnight I'm now starting to think that, at least as far as development goes, stopping + adding + starting is the way to go. My conclusion circles back to the first point: `stagedir`s + zip aren't a single virtual filesystem; trying to turn them into one would just be a very leaky abstraction.

That aside, this is my "one reliable way to hot reload Lua:"

 * give the Lua code a chance to serialize its state
 * destroy the old `lua_State`
 * create a new `lua_State`
 * re-run `.init.lua`
 * give the Lua code a chance to deserialize its saved state***

That is, I know, a rather limited view of "hot reloading." What it gets you is the ability to continue to listen for connections. The alternative is to stop redbean, swap the zip, start redbean — and stop listening while all of that is happening.

After pondering that overnight too, I think I'm OK with stop + swap + start in production. It's not ideal, but it's good enough, and I can do it with what's there now.

I don't think there's any way at all to do reliable hot reloading within a running Lua system; one module might hold references to things inside another module, or one module may have state that was build up by another module. A specific user might design a specific module to work within those limitations, but only as a special case. There's no way to generalize it.

===

***An even simpler version of that idea would be to just stop and start Lua without any serialization/deserialization of state. That's not unreasonable; web servers can stop any time anyway — so web apps have to make sure that important state is always serialized anyway. Without serialization/deserialization everything is much simpler — the model, the implementation, etc. But it's still additional complexity, and I'm not sure that complexity pays its own way.

**Comment by pkulchenko** [comment]
> But I get your bigger point. It's taken me some time and effort to work out the right mental model for what the stagedirs do/don't do.

I'd agree with that. Having said that, I wouldn't mind the package.path to be prepended, I just couldn't find a good way to both prepend and to keep the order in which paths in `ProgramDirectory` are specified. One idea is to mark a spot with double semicolons (`;;`), so if it's found, then insert the next one in front of it and keep it in the `package.path`. This way the paths get inserted in the correct order, but they are still before the "internal" zip paths. It also shouldn't impact anything else.

> That is, I know, a rather limited view of "hot reloading." What it gets you is the ability to continue to listen for connections. The alternative is to stop redbean, swap the zip, start redbean — and stop listening while all of that is happening.

I see two options for development: (1) stop and start and (2) ask redbean to reload its assets and files. I think serializing and killing the Lua state is a bit of an overkill, as it may be enough to trigger `OnServerReload` to allow your process to do whatever it wants to do. It may be to reload its templates, reload .init.lua, reload lua files or whatever else may be necessary. Simply reloading .init.lua may not work, as `package.loaded` table needs to be emptied, but overall it's not too difficult to do. This allows you to flexibly react to the reload event and potentially only reload specific functions or files depending on what you want to do. I actually have done something similar, but for a live coding from an IDE: http://notebook.kulchenko.com/zerobrane/live-coding-in-lua-bret-victor-style.

> After pondering that overnight too, I think I'm OK with stop + swap + start in production. It's not ideal, but it's good enough, and I can do it with what's there now.

For production I use a slightly different approach: I start a new redbean instance (after all the DB upgrades and checks are done) listening on the same port (using unix.SO_REUSEPORT) and send a termination signal to the "old" instance (based on a pid from its pidfile), which makes it stop accepting new connections and start draining the existing ones. The new instance starts accepting new connections as soon as it's launched, so there is no downtime. The process works quite well and allows a simple rollback to the previous version, although it does require launching a new instance with a new file name (which is usually not a problem).

**Comment by michaellenaghan** [comment]
> I just couldn't find a good way to both prepend and to keep the order in which paths in ProgramDirectory are specified...

Ah, I see, I hadn't thought of that. Here are two other ideas, to add to the one you already have:

 * Find the zip paths, remove them, append the new `stagedir`, append the zip paths
 * Find the zip paths, insert the new `stagedir` right before them

Just two slightly different ways of achieving the exact same thing.

> Simply reloading .init.lua may not work...

In addition to what you pointed out, another concern: redbean has its own rules and assumptions. A rule is that some things can only run in `.init.lua`. An assumption is that `.init.lua` will only run once. I don't know how hard it will be to get around those. We'll see!

>  I actually have done something similar, but for a live coding from an IDE: http://notebook.kulchenko.com/zerobrane/live-coding-in-lua-bret-victor-style.

That is very, very impressive. (I remember the Bret Victor talk.)

> For production I use a slightly different approach...

Thanks for that! Also very, very impressive. :-)

===

I've made some headway into creating a Docker image that creates executables I can run on my Mac. Creating the image and getting the build going were remarkably easy; I thought I'd be running into problems with missing dependencies, but `apt-get build-essentials` was pretty much all I needed.

Although I can run the executables that *were* produced, the build failed with an error while running tests:

```
error:test/libc/calls/fcntl_test.c:151: posixAdvisoryLocks_twoProcesses() on bf078f17e50c pid 55465 tid 55465
	ASSERT_EQ(-1, fcntl(3, F_SETLK, &lock))
		need -1	(or 0xffffffffffffffff) =
		 got 0	(or 0 or 
	EUNKNOWN/0/No error information
	o//test/libc/calls/fcntl_test.com @ bf078f17e50c
1 / 75 tests failed
```

I'll see if I can work out what's going on.

Thanks for all of your feedback and help, Paul.

**Comment by michaellenaghan** [comment]
> I'll see if I can work out what's going on.

I'm able to build `redbean.com`, etc. and run them on my Mac. I think I know why that particular test failed; I'll try to verify my theory later.

Fantastic.

I really thought, given what cosmopolitan does, that all of this would be *so* much harder…

**Comment by pkulchenko** [comment]
> Ah, I see, I hadn't thought of that. Here are two other ideas, to add to the one you already have:

Good ideas, although their drawback is that script changes to `package.path` would be overwritten and I'd expect redbean to be gentle about preserving `package.path` content and only adding what's needed without overwriting it.

> An assumption is that .init.lua will only run once. I don't know how hard it will be to get around those. 

That's a very good point! redbean provides and uses `OnlyCallFromInitLua`, but it also has `OnlyCallFromMainProcess`, which would check that the function is called from the main process (whether during the initial setup or later). We already ran into an issue with this for the `ProgramPrivateKey` call (https://github.com/jart/cosmopolitan/issues/816#issuecomment-1574629868), so I may need to review these calls to see which ones may need to be updated to use `OnlyCallFromMainProcess` instead.

> I'm able to build redbean.com, etc. and run them on my Mac. I think I know why that particular test failed; I'll try to verify my theory later.

That's great! I'm curious about the failing test though...

**Comment by michaellenaghan** [comment]
> Good ideas, although their drawback is that script changes to package.path would be overwritten and I'd expect redbean to be gentle about preserving package.path content and only adding what's needed without overwriting it.

Mmmm, no, that's not right. I'm suggesting: a) look for the zip paths; and b) insert the new `stagedir` before them. 

Let's say these are the zip paths:

    z1 z2

`cl1` and `cl2` are on the command-line, and processed in order. You end up first with:

    cl1 z1 z2

and then with:

    cl1 cl2 z1 z2

Now `.init.lua` runs, and the user modifies the path, prepending `u1` and appending `u2`:

    u1 cl1 cl2 z1 z2 u2

Now the user calls `ProgramDirectory()` with `pd1`:

    u1 cl1 cl2 pd1 z1 z2 u2

Now the user calls `ProgramDirectory()` with `pd2`:

    u1 cl1 cl2 pd1 pd2 z1 z2 u2

Just keep inserting before the zip paths.

Now, it's still possible for the user to really screw things up, but that's how it goes with shared state. I would say: just document what `ProgramDirectory()` does. If the user wants a different outcome they just have to call `ProgramDirectory()` before they manipulate the path themselves.

> I'm curious about the failing test though...

I was mounting the cosmo directory from my Mac into my Docker image:

    docker run -di --mount type=bind,source="$(shell pwd)/$(COSMO_DIR)",target=/opt/cosmo …

I do that with images I use for development, so that I can use Mac editors outside of Docker and immediately see the results inside of Docker. In this case, it also lets me run the built executables directly and immediately on my Mac — which is, you know, a first for me. :-) But apparently Docker's `mount` isn't fully Posix-compliant. No surprise there!

If I move the build tree into the Docker image, that test passes.

This one fails though:

```
error:test/libc/calls/getgroups_test.c:30: getgroups_test() on 402f9a427bb6 pid 1174 tid 1174
	EXPECT_EQ(0, setgroups(n, G))
		need 0	(or 0 or  =
		 got -1	(or 0xffffffffffffffff)
	EPERM/1/Operation not permitted
	o//test/libc/calls/getgroups_test.com @ 402f9a427bb6
error:test/libc/calls/getgroups_test.c:30: getgroups_test() on 402f9a427bb6 pid 1174 tid 1174
	EXPECT_EQ(0, EPERM/1/Operation not permitted)
		need 0	(or 0 or  =
		 got 1	(or 0x1 or '☺')
	EPERM/1/Operation not permitted
	o//test/libc/calls/getgroups_test.com @ 402f9a427bb6
2 / 2 tests failed
```

That seems to be related to [this](https://man7.org/linux/man-pages/man2/getgroups.2.html):

```
       setgroups() sets the supplementary group IDs for the calling
       process.  Appropriate privileges are required (see the
       description of the EPERM error, below).  The size argument
       specifies the number of supplementary group IDs in the buffer
       pointed to by list.  A process can drop all of its supplementary
       groups with the call:

           setgroups(0, NULL);

        …

ERRORS

        …

       EPERM  The calling process has insufficient privilege (the caller
              does not have the CAP_SETGID capability in the user
              namespace in which it resides).

       EPERM (since Linux 3.19)
              The use of setgroups() is denied in this user namespace.
              See the description of /proc/pid/setgroups in
              [user_namespaces(7)](https://man7.org/linux/man-pages/man7/user_namespaces.7.html).
```

I'm using `ubuntu:latest` as my base image, and running the build as root. Not quite sufficient for that particular test.

P.S. 

> I thought I'd be running into problems with missing dependencies, but `apt-get build-essentials` was pretty much all I needed.

For anyone who's interested: I overshot the mark on that. In fact, the [*only* build dependency](https://justine.lol/cosmopolitan/sources.html) is GNU `make`.

**Comment by pkulchenko** [comment]
Changed the order to prepending in #864.

**Comment by michaellenaghan** [comment]
Resolved in https://github.com/jart/cosmopolitan/commit/d967a94c9a6c9b830d91a90f5b0d34f4394ef9df
